rust/src/morton.rs

```diff
@@ -4,45 +4,43 @@ use {Box3, Result, Vec3};
 #[derive(PartialEq, Debug)]
 pub struct Morton(u64);
 
-fn shuffle(v: u64) -> u64 {
-    // take first 21 bits
-    let mut z = v & 0x00000000001fffff;
-    z = (z | (z << 32)) & 0x001f00000000ffff;
-    z = (z | (z << 16)) & 0x001f0000ff0000ff;
-    z = (z | (z << 8)) & 0x100f00f00f00f00f;
-    z = (z | (z << 4)) & 0x100f00f00f00f00f;
-    z = (z | (z << 2)) & 0x1249249249249249;
-
-    z
-}
-
-fn unshuffle(z: u64) -> u64 {
-    let mut v = z & 0x1249249249249249;
-    v = (v ^ (v >> 2)) & 0x10c30c30c30c30c3;
-    v = (v ^ (v >> 4)) & 0x100f00f00f00f00f;
-    v = (v ^ (v >> 8)) & 0x001f0000ff0000ff;
-    v = (v ^ (v >> 16)) & 0x001f00000000ffff;
-    v = (v ^ (v >> 32)) & 0x00000000001fffff;
-
-    v
-}
-
 impl<'a> From<&'a Vec3> for Morton {
     fn from(vec: &'a Vec3) -> Morton {
-        Morton(
-            (shuffle(vec.x as u64))
-                | (shuffle(vec.y as u64) << 1)
-                | (shuffle(vec.z as u64) << 2),
-        )
+        let x = vec.x as u64;
+        let y = vec.y as u64;
+        let z = vec.z as u64;
+        let mut idx = 0u64;
+        let bit_length = 64 - (std::cmp::max(x, std::cmp::max(y, z)) + 1).leading_zeros();
+        for i in 0..bit_length {
```

Comment:
Hmm, not a big fan of this. The code use to be constant-time. Now it's linear in the number of significant _bits_.

Do we understand the problem with the previous implementation? I'd prefer to fix that.

Reply:
I also wondered about this and researched a bit. I found this blog post https://www.forceflow.be/2013/10/07/morton-encodingdecoding-through-bit-interleaving-implementations/ which compares several approaches for implementing the morton encoding/decoding. Maybe the implementations in this diff were even based on that post? Either way, the post includes 3 (in my opinion) interesting details:

- in addition to the for-loop and to the magic-bit approach, there is also a lookup-table based approach (might be the best of both worlds?)
- the post has the disclaimer "The code used in this post is outdated and might have bugs." (which could explain why the old implementation of this library is buggy)
- the post refers to the C++ based `libmorton` library which is recommended (as it contains further improvements and fixes).

So, maybe it would be for the best to make use of the library? That would reduce the complexity on the wkw side, as well. Alternatively, we could try to gain insights from their sourcecode (this [file](https://github.com/Forceflow/libmorton/blob/main/libmorton/morton3D.h) would be most interesting probably). The relevant code functions seem to be rather short.

As another high-level thought: it's not really clear whether this code path is a bottleneck, is it? maybe it doesn't matter in the end. Nevertheless, using a prebaked solution sounds good to me.

Reply:
I think the perf-hit is negligible, since the Vec3 only has 32-bit components. That means the loop will be executed at max 32 times. In practice the loop will be called at max 5 times, bc the the bucket addresses within a file are at max 32. All operations in the loop are very cheap bit-operations on register values. The `leading_zeros` op is also a single CPU instruction.

> As another high-level thought: it's not really clear whether this code path is a bottleneck, is it?

My guess would be: absolutely not. Fetching a lookup table from uncached ram into register is probably slower than running this function.

I wouldn't want to pull in a lib (especially not a non-rust lib) for this due to the complexity at build time.

We could unroll the loop, but that would need to be dependent on `blocksPerFileDimLog2`.

Reply:
I implemented @amotta's fix in https://github.com/scalableminds/webknossos-wrap/compare/lz4-crate...re-morton. Seems to work fine.